ts/private/ts_project_worker.js

```diff
@@ -0,0 +1,384 @@
+        getDirectories: getDirectories,
+        watchFile: watchFile,
+        watchDirectory: watchDirectory
+    }
+
+    const host = ts.createWatchCompilerHost(
+        tsconfig,
+        parsedArgs.options,
+        sys,
+        createEmitCacheAndDiagnosticsProgram,
+        noop,
+        noop
+    )
+
+    host.invalidate = invalidate;
+
+    program = ts.createWatchProgram(host)
+    
+    function getDirectoryWatcherForPath(path) {
+        for (const [directory, cb] of directoryWatchers.entries()) {
+            if (isAncestorDirectory(directory, path)) {
+                debuglog(`found a directory watcher for ${path} ${directory}`)
+                return cb;
+            }
+        }
+    }
+
+    function writeFile(path, data, writeByteOrderMark) {
+        debuglog(`writeFile ${path}`);
+        ts.sys.writeFile(path, data, writeByteOrderMark);
+    }
+
+    function readFile(filePath, encoding) {
+        debuglog(`readFile ${filePath}`);
+        const relative = path.relative(execRoot, filePath)
+        /** if it is under node_modules just allow file reads as we don't have a list of deps */
+        if (filePath.indexOf("node_modules") == -1 && !knownInputs.has(relative)) {
+            return undefined
+        }
+        return ts.sys.readFile(filePath, encoding);
+    }
+
+    function directoryExists(directory) {
+        if (!directory.startsWith(execRoot)) {
+            return false;
+        }
+        const exists = ts.sys.directoryExists(directory)
+        directory.indexOf("node_modules") == -1 && debuglog(`directoryExists ${directory} ${exists}`);
+        return exists
+    }
+
+    function getDirectories(directory) {
+        const dirs = ts.sys.getDirectories(directory);
+        debuglog(`getDirectories ${directory}`, dirs)
+        return dirs;
+    }
+
+    function fileExists(filePath) {
+        const relative = path.relative(execRoot, filePath)
+        debuglog(`fileExists ${filePath} ${knownInputs.has(relative)}`)
+        if (!knownInputs.has(relative)) {
+            return false;
+        }
+        return true;
+    }
+
+    function readDirectory(directory, extensions, exclude, include, depth) {
+        const files = ts.sys.readDirectory(directory, extensions, exclude, include, depth);
+        const strictView = [];
+
+        for (const file of files) {
+            const relativePath = path.relative(execRoot, file);
+            if (knownInputs.has(relativePath)) {
+                strictView.push(file);
+            } else {
+                debuglog(`Skipping ${relativePath} as it is not input to current compilation.`)
+            }
+        }
+    
+
+        debuglog(`readDirectory ${directory} ${strictView}`)
+
+        return strictView;
+    }
+
+    function invalidate(filePath, kind) {
+        if (filePath.endsWith(".params")) return;
+        debuglog(`invalidate ${filePath} : ${ts.FileWatcherEventKind[kind]}`)
+
+        if (kind == ts.FileWatcherEventKind.Created) { 
```

Comment:
Normally `===` is recommended everywhere. Can we just do a search and replace and always use the triple-equals?

Reply:
Done. sometimes I hate javascript so much.